File: Data Structures & Algorithms/subsets/submission-1.go
// add: append(s, e)
// pop: s = s[:len(s)-1]



func subsets(nums []int) [][]int {
	res := [][]int{}
	
	var dfs func(curr []int, i int)
	dfs = func(curr[]int, i int) {
		if i >= len(nums) {
			snapshot := make([]int, len(curr))
			copy(snapshot, curr)
			res = append(res, snapshot)
			return
		}
		// add the current element and go next
		curr = append(curr, nums[i])
		dfs(curr, i+1)

		// pop the last element and go next
		curr = curr[:len(curr)-1]
		dfs(curr, i+1)
	}
	dfs([]int{}, 0)
	return res
}

File: Data Structures & Algorithms/permutations/submission-1.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        
        res = []

        def dfs(candidates, curr):
            if not candidates: # length of candidates is 0
                res.append(curr)
                return

            
            for i in range(len(candidates)):
                new_curr = list.copy(curr)
                new_curr.append(candidates[i])
                dfs(candidates[:i] + candidates[i+1:], new_curr)

        dfs(nums, [])
        return res

            
        